### PR TITLE
Sort numeric ref labels numerically

### LINKED ISSUE #314: Use human numeric sorting
### Describe the bug

**context**
When sorting footnotes, mdformat produces this diff:

```diff
diff --git i/rules.md w/rules.md
index 0f7f85d..492b320 100644
--- i/rules.md
+++ w/rules.md
@@ -654,6 +654,7 @@ trademarks are property of their respective owners.

 <!-- EXLINKS -->

+[10]: https://sirplease.gg/index.php?p=banlist
 [3]: https://en.wikipedia.org/wiki/Swiss-system_tournament
 [4]: http://www.fide.com/component/handbook/?id=84&view=article
 [5]: http://www.wizards.com/dci/downloads/Swiss_Pairings.pdf
@@ -661,4 +662,3 @@ trademarks are property of their respective owners.
 [7]: https://gist.github.com/rpdelaney/1b88b128b76b1e1a71fec59d0b5ffb14
 [8]: https://magic.wizards.com/en/articles/archive/cumulative-tiebreak-2013-08-01
 [9]: https://senseis.xmp.net/?SonnebornBerger
-[10]: https://sirplease.gg/index.php?p=banlist
```

**expectation**
I expected `10` to sort after `9`.

**bug**
But instead `10` sorts before `3`.

**problem**
This is a problem because humans expect bigger numbers to sort after smaller ones.

### Reproduce the bug

Run mdformat on this file:

```markdown
(Lorem)[3] (ipsum)[4] (dolor)[5] (sit)[6] (amet,)[7] (consectetur)[8] (adipiscing)[9] (elit,)[10]...

[3]: https://en.wikipedia.org/wiki/Swiss-system_tournament
[4]: http://www.fide.com/component/handbook/?id=84&view=article
[5]: http://www.wizards.com/dci/downloads/Swiss_Pairings.pdf
[6]: https://github.com/SirPlease/IPTables
[7]: https://gist.github.com/rpdelaney/1b88b128b76b1e1a71fec59d0b5ffb14
[8]: https://magic.wizards.com/en/articles/archive/cumulative-tiebreak-2013-08-01
[9]: https://senseis.xmp.net/?SonnebornBerger
[10]: https://sirplease.gg/index.php?p=banlist
```

### List your environment

```console
$ mdformat --version
mdformat 0.7.13
$ python3 --version
Python 3.10.0
```

### PR DESCRIPTION
**Pull request by rpdelaney**
Closes #314

### CODE REVIEW

**Review by hukkin — CHANGES_REQUESTED** [review]

File: tests/test_style.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "fixture_file,options",
     [
         ("default_style.md", []),
+        ("numeric_references.md", []),
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply tests/test_style.py]: bd70c75

File: src/mdformat/renderer/__init__.py [hunk 2]
```diff
@@ -101,7 +101,10 @@ def render_tree(
     @staticmethod
     def _write_references(env: MutableMapping) -> str:
         ref_list = []
-        for label in sorted(env["used_refs"]):
+        for label in sorted(
+            env["used_refs"],
+            key=lambda x: (not x.isnumeric(), int(x) if x.isnumeric() else x),
+        ):
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply src/mdformat/renderer/__init__.py:107]: > str.isnumeric allows characters other than ASCII digits, which will make the int that follows raise an error

True. I've added `⅕` as a test case and I'll work on extending the solution to that.

> numeric labels were sorted above everything else, including special characters. Not sure we want that.

I'd think they have to sort either before or after. I'm not aware of a collation that sorts numerics interlaced with non-numerics. I don't have a strong opinion about which should go first, though: numerics first was practically the outcome of a coin toss. If you prefer, I can easily reverse that.

[the remark this answers is not in the document]
Comment (hukkin) [reply src/mdformat/renderer/__init__.py:107]: > True. I've added ⅕ as a test case and I'll work on extending the solution to that.

Thanks for adding the test. Not sure what you mean by extending the solution, however. I think this PR should only deal with ASCII digits. The suggested solution should do exactly that.

[the remark this answers is not in the document]
Comment (rpdelaney) [reply src/mdformat/renderer/__init__.py:107]: I added some more corner cases to the test. Let me know how you think they should sort:

```
+[1]: .
+[2]: .
+[10]: .
+[word]: .
+[⁵]: .
+[⅕]: .
+[𐩂]: .
+[🤪]: .
```

[the remark this answers is not in the document]
Comment (hukkin) [reply src/mdformat/renderer/__init__.py:107]: Haha, there's probably a gazillion things we could consider: Chinese numerals, Japanese numerals, comma vs period as a separator, fractions etc.

This is why I ask that we only consider labels that are ASCII digits only. It should fix your use case, and we can consider other use cases if people that use them start asking.

[the remark this answers is not in the document]
Comment (rpdelaney) [reply src/mdformat/renderer/__init__.py:107]: Yeah, you're right. This is way too big a problem space to solve everything at once. Some numeral systems in use today in natural language aren't base 10. How should they sort alongside Arabic digits? Perhaps `mdformat` should convert [Katovik numerals](https://en.wikipedia.org/wiki/Kaktovik_numerals) to Hindu-Arabic before sorting them in? What about non-positional numeral systems? Should `CDXXV` sort before or after `426`?

I'll come back to this a bit later lol

**Comment by rpdelaney** [comment]
@hukkin How's that?

File: src/mdformat/renderer/__init__.py [hunk 3]
```diff
@@ -100,8 +101,16 @@ def render_tree(
 
     @staticmethod
     def _write_references(env: MutableMapping) -> str:
+        def label_sort_key(label: str) -> str:
+            if not label:
+                raise AssertionError("link label cannot be empty string")
```
Comment (hukkin) [inline src/mdformat/renderer/__init__.py:106]:
```suggestion
            assert label, "link label cannot be empty string"
```
This is better because it get optimized away with `python -O`

File: src/mdformat/renderer/__init__.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
 from collections.abc import Mapping, MutableMapping, Sequence
 import logging
+from string import digits
```
Comment (hukkin) [inline src/mdformat/renderer/__init__.py:14]:
```suggestion
import string
```

File: src/mdformat/renderer/__init__.py [hunk 5]
```diff
@@ -100,8 +101,16 @@ def render_tree(
 
     @staticmethod
     def _write_references(env: MutableMapping) -> str:
+        def label_sort_key(label: str) -> str:
+            if not label:
+                raise AssertionError("link label cannot be empty string")
+            if all(c in digits for c in label):
```
Comment (hukkin) [inline src/mdformat/renderer/__init__.py:107]:
```suggestion
            if all(c in string.digits for c in label):
```
I prefer to keep the stdlib context here.

**Comment by hukkin** [comment]
Merged, thanks for your contribution!

File: src/mdformat/renderer/__init__.py (hunk 3 above)
Comment (rpdelaney) [reply src/mdformat/renderer/__init__.py]: Yep! I try to avoid using `assert` for runtime validation for that reason.

Comment (hukkin) [reply src/mdformat/renderer/__init__.py]: IMO it's desirable here because the `if not label` condition is a something that should never happen unless there's a programming error. A label cannot be an empty string in CommonMark. The check is mostly there for peace of mind and to prove the programmer that the `if not label` condition is never true (if it were, it'd have implications for the `all()` calll that follows).
